### ISSUE TITLE
Style/RedundantRegexpEscape generates false positives for a hyphen (-) inside a character group ([ ])

### ISSUE THREAD

**Issue by JoeCohen** [rubocop/rubocop#10430]
## Is your feature request related to a problem? Please describe.
Style/RedundantRegexpEscape generates false positives for an escaped hyphen (-) inside a character group when the user intends to match a literal hyphen. 
In Ruby a hyphen (-) inside a regex character group is a metacharacter. https://ruby-doc.org/core-3.1.0/Regexp.html#class-Regexp-label-Character+Classes. Quoting or not quoting it yields different results. See code example below.
Therefore:
- If the user intends to use it as a metacharacter, it must not be quoted.
- If the user intends to match a literal hyphen, it must be quoted.

```ruby
> bee = "b"
 => "b" 
> hyphen = "-"
 => "-" 
> unescaped = /[a-c]/
 => /[a-c]/ 
> escaped = /[a\-c]/
 => /[a\-c]/ 
> unescaped.match?(bee)
 => true 
> unescaped.match?(hyphen)
 => false 
> escaped.match?(bee)
 => false 
> escaped.match?(hyphen)
 => true 
```

I don't think there's a way to determine what the user intended.

## Describe the solution you'd like
Please consider:
-  designating Style/RedundantRegexpEscape **not** Safe, and/or
-  generating a Warning for both quoted and unquoted hyphens inside a regex character group.

## Related Issue
#8623 Bug with using - inside brackets [] with using Style/RedundantRegexpEscape

**Comment by dvandersluis** [comment]
With the following code:

```
unescaped = /[a-c]/
escaped = /[a\-c]/
```

`Style/RedundantRegexpEscape` does not register any offenses for either regexp. I'm going to close this issue, but if you have a failing case (please make sure your RuboCop is up to date) please feel free to reopen!

**Comment by andretges** [comment]
Not exactly the same issue, but similar one. En dash `–` accidentally used and escaped in the regexp crashes RuboCop:
```ruby
> "–".ord # en dash
=> 8211

FORMAT = /[a\–c]/
```

```sh
[~/code/project]$ bundle exec rubocop -d app/models/client.rb
For /Users/andrei/code/project: configuration from /Users/andrei/code/project/.rubocop.yml
Inheriting configuration from /Users/andrei/code/project/.rubocop-shared.yml
configuration from /Users/andrei/code/project/.rubocop-https---raw-githubusercontent-com-project-rubocop-project-v1-50-0--rubocop-shared-yml
Default configuration from /Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/config/default.yml
Inheriting configuration from /Users/andrei/code/project/.rubocop-rails.yml
configuration from /Users/andrei/code/project/.rubocop-https---raw-githubusercontent-com-project-rubocop-project-v1-50-0--rubocop-rails-yml
configuration from /Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-rails-2.21.0/config/default.yml
configuration from /Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-rails-2.21.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/andrei/code/project/app/models/client.rb
An error occurred while Style/RedundantRegexpEscape cop was inspecting /Users/andrei/code/project/app/models/client.rb:26:11.
invalid byte sequence in UTF-8
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:64:in `match?'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:64:in `allowed_escape?'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:47:in `block in on_regexp'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:100:in `block in each_escape'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:39:in `block in traverse'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/subexpression.rb:28:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/subexpression.rb:28:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:37:in `each_with_index'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:37:in `traverse'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:42:in `block in traverse'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/subexpression.rb:28:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/subexpression.rb:28:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:37:in `each_with_index'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/regexp_parser-2.8.1/lib/regexp_parser/expression/methods/traverse.rb:37:in `traverse'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:99:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:99:in `reduce'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:99:in `each_escape'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/style/redundant_regexp_escape.rb:46:in `on_regexp'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:107:in `public_send'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:107:in `block (2 levels) in trigger_responding_cops'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:171:in `with_cop_error_handling'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:106:in `block in trigger_responding_cops'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:105:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:105:in `trigger_responding_cops'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:69:in `on_regexp'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:152:in `on_casgn'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:71:in `on_casgn'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:137:in `block in on_dstr'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:137:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:137:in `on_dstr'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:71:in `on_begin'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:154:in `on_class'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:71:in `on_class'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-ast-1.29.0/lib/rubocop/ast/traversal.rb:20:in `walk'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/commissioner.rb:87:in `investigate'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/team.rb:156:in `investigate_partial'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cop/team.rb:98:in `investigate'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:345:in `block in inspect_file'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:344:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:344:in `flat_map'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:344:in `inspect_file'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:287:in `block in do_inspection_loop'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:321:in `block in iterate_until_no_changes'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:314:in `loop'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:314:in `iterate_until_no_changes'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:283:in `do_inspection_loop'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:164:in `block in file_offenses'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:189:in `file_offense_cache'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:163:in `file_offenses'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:154:in `process_file'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:135:in `block in each_inspected_file'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:134:in `each'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:134:in `reduce'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:134:in `each_inspected_file'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:120:in `inspect_files'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/runner.rb:73:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/command.rb:11:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli/environment.rb:18:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli.rb:118:in `run_command'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli.rb:125:in `execute_runners'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli.rb:51:in `block in run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli.rb:77:in `profile_if_needed'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/lib/rubocop/cli.rb:43:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/exe/rubocop:19:in `block in <top (required)>'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/3.2.0/benchmark.rb:311:in `realtime'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.50.2/exe/rubocop:19:in `<top (required)>'
/Users/andrei/.rbenv/versions/3.2.2/bin/rubocop:25:in `load'
/Users/andrei/.rbenv/versions/3.2.2/bin/rubocop:25:in `<top (required)>'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli/exec.rb:58:in `load'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli/exec.rb:58:in `kernel_load'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli/exec.rb:23:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli.rb:492:in `exec'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli.rb:34:in `dispatch'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/cli.rb:28:in `start'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/exe/bundle:37:in `block in <top (required)>'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
/Users/andrei/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.14/exe/bundle:29:in `<top (required)>'
/Users/andrei/.rbenv/versions/3.2.2/bin/bundle:25:in `load'
/Users/andrei/.rbenv/versions/3.2.2/bin/bundle:25:in `<main>'
.

1 file inspected, no offenses detected

1 error occurred:
An error occurred while Style/RedundantRegexpEscape cop was inspecting /Users/andrei/code/project/app/models/client.rb:26:11.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
https://github.com/rubocop/rubocop/issues

Mention the following information in the issue report:
1.50.2 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [x86_64-darwin22]
Finished in 2.846841999795288 seconds
```

**Comment by ydakuka** [comment]
I tried to reproduce and couldn't.

Ruby:
```console
ydakuka@yauhenid:~/Work/project$ ruby -v
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-linux]
```

I have the code:
```ruby
class Test
  FORMAT_ESCAPED = /[a\–c]/.freeze
  FORMAT_UNESCAPED = /[a–c]/.freeze
end
```

I run the command `rubocop app/models/test.rb` and don't get any offences.
```console
ydakuka@yauhenid:~/Work/project$ rubocop app/models/test.rb  --only Style/RedundantRegexpEscape
Inspecting 1 file
.

1 file inspected, no offenses detected
```
